File: dicm2dckv/dicm2cda/dckvapi.m
//
//  dckvapi.m
//  dicm2dckv
//
//  Created by jacquesfauquex on 2024-04-04.
//

/*potencialmente multivalue: AE AS AT CS DA DS DT FD FL IS LO OW PN SH SV TM UC UI UV*/
#include <Foundation/Foundation.h>

#include "dckvapi.h"

FILE *outFile;
static char *dbpath;

//possibility to overwrite any value read
size_t dckvapi_fread(
                     void * __restrict __ptr,
                     size_t __size,
                     size_t __nitems,
                     FILE * __restrict __stream
                     )
{
   return fread(__ptr,__size,__nitems,__stream);
}

u8 swapchar;


//returns true when 8 bytes were read
//possibility to overwrite any tag,vr,length reading
BOOL dckvapi_fread8(uint8_t *buffer, u64 *bytesReadRef)
{
   *bytesReadRef=fread(buffer, 1, 8, stdin);
   if (ferror(stdin)){
      E("%s","stdin error");
      return false;
   }
   
   if (*bytesReadRef==8){
      swapchar=*buffer;
      *buffer=*(buffer+1);
      *(buffer+1)=swapchar;
      swapchar=*(buffer+2);
      *(buffer+2)=*(buffer+3);
      *(buffer+3)=swapchar;
   }
   else
   {
      *buffer=0xFF;
      *(buffer+1)=0xFF;
      *(buffer+2)=0xFF;
      *(buffer+3)=0xFF;
   }
   return true;
}

const char *space=" ";
const char *backslash = "\\";

#pragma mark obligatorios formales
bool createtx(
   const char * dstdir,
   uint8_t    * vbuf,
   u64 *soloc,         // offset in valbyes for sop class
   u16 *solen,         // length in valbyes for sop class
   u16 *soidx,         // index in const char *scstr[]
   u64 *siloc,         // offset in valbyes for sop instance uid
   u16 *silen,         // length in valbyes for sop instance uid
   u64 *stloc,         // offset in valbyes for transfer syntax
   u16 *stlen,         // length in valbyes for transfer syntax
   u16 *stidx,         // index in const char *csstr[]
   s16 *siidx          // SOPinstance index
){
   (*siidx)++;
   D("#%d",*siidx);
   
   dbpath=malloc(0xFF);
   strcat(dbpath,dstdir);
   strcat(dbpath, "/");
   char *ibuf = malloc(*silen);
   memcpy(ibuf, vbuf+*siloc+8, *silen);
   strcat(dbpath,ibuf);
   strcat(dbpath, ".dscd.xml");
   outFile=fopen(dbpath, "w");

   return true;
}
bool committx(s16 *siidx){
   return closetx(siidx);
}
bool closetx(s16 *siidx){
   fclose(outFile);
   return true;
}


#pragma mark - parseo y agregado

bool appendkv(
              uint8_t           *kbuf,
              u32                kloc,
              bool               vlenisl,
              enum kvVRcategory  vrcat,
              u64                vloc,
              u32                vlen,
              bool               fromStdin,
              uint8_t           *vbuf
              )
{
   D("%d\n",vrcat);

   switch (vrcat) {
         
      /*
      case kvdn://ST  DocumentTitle 00420010
      {
         if (vlen && (fread(vbuf,1,vlen,stdin)!=vlen)) return false;
         printf("%s","title");
         if (vlen > 0)
         {
            printf(" (");
            vbuf[vlen]=0x0;//terminate even uid lists
            char *token = strtok((char *)vbuf, backslash);
            while( token != NULL ) {
               printf( " «%s»", token );
               token = strtok(NULL, backslash);
            }
            printf(" )");
         }
         printf("\n");
      } break;
      */
         
      case kved://OB encapsulaed document
      {
         printf("%s","doc\n");
         if (fromStdin && vlen)
         {
            u32 toberead=vlen;
            while (toberead>0xFFFE)
            {
               if (fread(vbuf,1,0xFFFE,stdin)!=0xFFFE) return false;
               toberead-=0xFFFE;
               if ((toberead==0)&&(vbuf[0xFFFD]==0x0)&&(fwrite(vbuf ,1, 0xFFFD , outFile)!=0xFFFD));
               else if (fwrite(vbuf ,1, 0xFFFE , outFile)!=0xFFFE) return false;
            }
            if (toberead)
            {
               if ((!feof(stdin)) && (fread(vbuf,1,toberead,stdin)!=toberead)) return false;
               if (vbuf[toberead-1]==0x0) toberead--;
               if (fwrite(vbuf ,1, toberead , outFile)!=toberead) return false;
            }
         }
          
      } break;
      
      default:
      {
         if (fromStdin && vlen)
         {
            while (vlen>0xFFFE)
            {
               if (fread(vbuf,1,0xFFFE,stdin)!=0xFFFE) return false;
               vlen-=0xFFFE;
            }
            if (vlen && (!feof(stdin)) && (fread(vbuf,1,vlen,stdin)!=vlen)) return false;
         }
      }
   }

   return true;
}

File: dicm2dckv/dicm2cda/dckvapi.c
// project: dicm2dckv
// target: dicm2cda
// file: dckvapi.c
// created by jacquesfauquex on 2024-04-04.

#include "dckvapi.h"

extern char *DICMbuf;
extern u64 DICMidx;
extern s16 siidx;
extern uint8_t *kbuf;

#pragma mark - read

static u64 bytesreceived;
bool _DKVfread(u32 bytesaskedfor)
{
   bytesreceived=fread(DICMbuf+DICMidx,1,bytesaskedfor,stdin);
   if (bytesreceived>0xFFFFFFFF)return 0;
   DICMidx+=bytesreceived;
   return (bytesaskedfor==bytesreceived);
}

//returns true when 8(+4) bytes were read
bool _DKVfreadattr(u8 kloc)
{
   if (fread(DICMbuf+DICMidx,1,8,stdin)!=8)
   {
      if (ferror(stdin)) D("%s","stdin error");
      return false;
   }
   
   //group LE>BE
   kbuf[kloc]=DICMbuf[DICMidx+1];
   kbuf[kloc+1]=DICMbuf[DICMidx];
   //element LE>BE
   kbuf[kloc+2]=DICMbuf[DICMidx+3];
   kbuf[kloc+3]=DICMbuf[DICMidx+2];
   //vr vl copied (LE)
   kbuf[kloc+4]=DICMbuf[DICMidx+4];
   kbuf[kloc+5]=DICMbuf[DICMidx+5];
   kbuf[kloc+6]=DICMbuf[DICMidx+6];
   kbuf[kloc+7]=DICMbuf[DICMidx+7];

   switch ((DICMbuf[DICMidx+5]<<8)|(DICMbuf[DICMidx+4])) {
      case OB://other byte
      case OW://other word
      case OD://other double
      case OF://other float
      case OL://other long
      case SV://signed 64-bit very long
      case OV://other 64-bit very long
      case UV://unsigned 64-bit very long
      case UC://unlimited characters
      case UT://unlimited text
      case UR://universal resrcurl identifier/locator
      case SQ://sequence
      {
         DICMidx+=8;
         if (fread(DICMbuf+DICMidx,1,4,stdin)!=4)
         {
            if (ferror(stdin)) E("%s","stdin error");
            return false;
         }
         memcpy(kbuf+kloc+8, DICMbuf+DICMidx, 4);
         DICMidx+=4;
      }break;
      default:
      {
         //IA,IZ,SZ require postprocessing in dicm2dckv
         DICMidx+=8;
         memcpy(kbuf+kloc+8, DICMbuf+DICMidx-2, 2);
         kbuf[kloc+10]=0;
         kbuf[kloc+11]=0;
      }break;
   }
   
   return true;
}

#pragma mark - instance transactions
static char *dbpath;
static FILE *outFile;

bool _DKVcreate(
   u64 soloc,         // offset in valbyes for sop class
   u16 solen,         // length in valbyes for sop class
   u16 soidx,         // index in const char *scstr[]
   u64 siloc,         // offset in valbyes for sop instance uid
   u16 silen,         // length in valbyes for sop instance uid
   u64 stloc,         // offset in valbyes for transfer syntax
   u16 stlen,         // length in valbyes for transfer syntax
   u16 stidx          // index in const char *csstr[]
){
   dbpath=malloc(0xFF);
   char *ibuf = malloc(silen);
   memcpy(ibuf, DICMbuf+siloc, silen);
   strcat(dbpath,ibuf);
   strcat(dbpath, ".dscd.xml");
   outFile=fopen(dbpath, "w");
   return true;
}

bool _DKVclose(void){
   fclose(outFile);
   return true;
}

static u32 titlerepidx=0;
static u32 titleoffset=0;
static u32 titlelength=0;
static u32 documentoffset=0;
static u32 documentlength=0;
bool _DKVcommit(bool hastrailing){
   //title charset -> utf-8
   u32 utf8length=0;
   utf8(titlerepidx,DICMbuf,titleoffset,titlelength,DICMbuf,(u32)DICMidx,&utf8length);
   printf( "%.*s\n", utf8length,DICMbuf+DICMidx );
   
   //write document
   if (!fwrite(DICMbuf+documentoffset ,1, documentlength , outFile)) return false;

   return _DKVclose();
}

#pragma mark - write

bool _DKVappend(int kloc, enum kvVRcategory  vrcat, u32 vlen)
{
   switch (vrcat) {
      case kvSA:
      case kvSZ:
      case kvIA:
      case kvIZ: break;
      case kvsdoctitle: { //ST  DocumentTitle 00420010
         titlerepidx=kbuf[kloc+6] + (kbuf[kloc+7] << 8);
         titleoffset=(u32)DICMidx;
         titlelength=vlen;
         if ((vlen > 0) && (!_DKVfread(vlen))) return false;
      } break;
      case kvsdocument: { //OB encapsulaed document
         if (!_DKVfread(vlen)) return false;
         documentoffset=(u32)(DICMidx-vlen);
         documentlength=vlen - (DICMbuf[DICMidx-1]==0);//last char 0x00 ?
      } break;
      default:if (!_DKVfread(vlen)) return false;break;
   }
   return true;
}
